presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkOutputOperator.java

```diff
@@ -0,0 +1,295 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spark.execution;
+
+import com.facebook.presto.execution.buffer.PagesSerdeFactory;
+import com.facebook.presto.operator.DriverContext;
+import com.facebook.presto.operator.Operator;
+import com.facebook.presto.operator.OperatorContext;
+import com.facebook.presto.operator.OperatorFactory;
+import com.facebook.presto.operator.OutputFactory;
+import com.facebook.presto.operator.PartitionFunction;
+import com.facebook.presto.spark.classloader_interface.PrestoSparkRow;
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.RunLengthEncodedBlock;
+import com.facebook.presto.spi.plan.PlanNodeId;
+import com.facebook.presto.spi.relation.ConstantExpression;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.OutputPartitioning;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.slice.SliceOutput;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.function.Function;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.lang.Integer.min;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+public class PrestoSparkOutputOperator
+        implements Operator
+{
+    public static class PrestoSparkOutputFactory
+            implements OutputFactory
+    {
+        private static final OutputPartitioning SINGLE_PARTITION = new OutputPartitioning(
+                new ConstantPartitionFunction(),
+                ImmutableList.of(),
+                ImmutableList.of(),
+                false,
+                OptionalInt.empty());
+
+        private final PrestoSparkRowBuffer rowBuffer;
+
+        public PrestoSparkOutputFactory(PrestoSparkRowBuffer rowBuffer)
+        {
+            this.rowBuffer = requireNonNull(rowBuffer, "rowBuffer is null");
+        }
+
+        @Override
+        public OperatorFactory createOutputOperator(
+                int operatorId,
+                PlanNodeId planNodeId,
+                List<Type> types,
+                Function<Page, Page> pagePreprocessor,
+                Optional<OutputPartitioning> outputPartitioning,
+                PagesSerdeFactory serdeFactory)
+        {
+            OutputPartitioning partitioning = outputPartitioning.orElse(SINGLE_PARTITION);
```

Comment:
> For example, in

Github must've seved not the final version. Could you pleas elaborate?

Reply:
@arhimondr : @arhimondr : For example, in `TaskOutputOperatorFactory` ,  the output partitioning can be any of the `FIXED_BROADCAST_DISTRIBUTION`, `FIXED_ARBITRARY_DISTRIBUTION`, `SCALED_WRITER_DISTRIBUTION`, `SINGLE_DISTRIBUTION`, `COORDINATOR_DISTRIBUTION`: https://github.com/prestodb/presto/blob/2fb54ed0dc380791876810237c116f587efc2470/presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java#L381-L395

Update: with the check argument that output partitioning must be absent, it might be OK... just note the absence of output partitioning seems to have different semantic for `TaskOutputOperatorFactory` vs. `PrestoSparkOutputFactory`